docs/structured-streaming-programming-guide.md

```diff
@@ -3146,6 +3146,17 @@ See [Input Sources](#input-sources) and [Output Sinks](#output-sinks) sections f
       - After `coalesce`, the number of (reduced) tasks will be kept unless another shuffle happens.
   - `spark.sql.streaming.stateStore.providerClass`: To read the previous state of the query properly, the class of state store provider should be unchanged.
   - `spark.sql.streaming.multipleWatermarkPolicy`: Modification of this would lead inconsistent watermark value when query contains multiple watermarks, hence the policy should be unchanged.
+- Structured Streaming uses `global watermark` which might impact on query having multiple stateful operations.
+  - Stateful operators: aggregation, deduplication, stream-stream join, (flat)mapGroupsWithState
+  - You should be able to answer below questions for your query to get correct outputs:
+    - How global watermark is calculated on your query?
+    - How global watermark is applied to each stateful operator?
+    - Is there any intermediate output being discarded as "late input" due to watermark?
+  - Fail to answer above questions might lead to incorrect outputs - e.g. intermediate outputs being discarded. 
```

Comment:
This is probably an important topic but I'm not clear what the guidance is here. If my answers are what then what do I do? what do you mean fail to answer the question? can this be more direct? what's an example of a problem, cause and solution?

Reply:
https://issues.apache.org/jira/browse/SPARK-28094 is the example on problem, as I tried to help resolving that issue (in user mailing list) and found the correctness issue.

> If my answers are what then what do I do?

The set of questions are asking theirselves as whether they're fully understanding the details of watermark, because it could bring "unexpected" discard on rows between stateful operators, say, correctness issue.

So it's like an exam to prevent end users to "just do it as Spark allows it" without fully understanding what they're doing. If they can't answer with their query, they should find alternative approach.

> what do you mean fail to answer the question?

Literally. If they are having any small unsure thing about answering these question they fail. I know it's too hard on end users, but well, better than realizing incorrect outputs in production env.

> can this be more direct? what's an example of a problem, cause and solution?

I'm happy to do it but not sure how much I need to explain with details. If we want to have detailed explanation like providing example, maybe better to have individual chapter. It might be  long enough to use as a blog post. This is an advanced topic - the book "Streaming Systems" assigns a chapter for watermark - so may not be feasible to add the details to the guide page, but due to the importance we should mention it. So that looks to be a dilemma.

To be honest, there's no "solution", only workaround, since it's due to lack of feature - watermark propagation and stateful operator-wise watermark. https://issues.apache.org/jira/browse/SPARK-26655 tries to address this, but not much efforts are here due to lack of interests.

Reply:
Yes, this might be too much for the docs here, but can we explain more clearly what type of incorrect usage may cause what problem? so that users know what to look for, in both their causes and symptoms. Otherwise it's just kind of saying "there are problems here if you're not careful".

Reply:
> what type of incorrect usage may cause what problem

Incorrect usages can't be collected out as some types - Spark allows arbitrary operations on state via map/flatMapGroupsWithState, so can't enumerate the overall usages and picks up incorrect usages. Users are open to encounter the situation whenever they're using multiple stateful operations in a query, according to their query.

One symptom would be incorrect outputs as SPARK-26655 represents, but Spark doesn't provide any information to indicate the issue so I had to analyze the query via answering these questions for one of query in SPARK-26655.

Maybe we shouldn't refer the case as "incorrect usage" even they've encountered, since the issue is coming from design and limitation on watermark in structured streaming. It's not their fault and they're not doing wrong: Spark just doesn't support it.

As I commented below, maybe disallowing the cases (with guiding alternative) until we make it correct would be better approach, but I'd defer to committers/PMCs on how to handle this.